### PR TITLE
Fixes #211 Check SAME_AS_CONTAINER for container write access

### LINKED ISSUE #211: Permssion Handling Definition contained by Requirement on Update
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the COMET Web Services
- [x] I have searched [open](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
As a Participant that has MODIFY_IF_OWNED on  Requirement i was able to modify the contained definition of a non-owned requirement, seems something may be wrong with the permission handler. 

The SiteDirectory does not contain an explicit permission for the Definition class.

**_07-07-21 16:10 Edit Alexander:_**
Problem is caused by the OperationProcessor.ApplyUpdateOperations method.
The `RequestSecurityContext`'s ContainerReadAllowed and ContainerWriteAllowed properties are set to `true` by default.
The PermissionService returns the value of the ContainerWriteAllowed property when a Thing's ParticipantAccessRight is set to SAME_AS_CONTAINER, which is the case for a Definition object.
The solution should be to get the Thing's container from the operationThingCache if it's there and otherwise get it from the database and run the CanWrite for the Container thing. Could be that extra recursion, with performance degradation  could take place here.

### Steps to Reproduce
using the "specification browser" edit a requirement that is not owned; this may or may not be possible depending on the state of issue https://github.com/RHEAGROUP/COMET-IME-Community-Edition/issues/838

### System Configuration
- COMET Web Services version: 6.1.4
  - [x] CDP4Common:  7.3.0
  - [x] PostrgreSQL: 12.7-1

### PR DESCRIPTION
**Pull request by lxatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WebServices-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
Check SAME_AS_CONTAINER for container write access

### CODE REVIEW

**Comment by alexatrhea** [comment]
So how do you resolve the container now? Not obvious from the code.

File: CDP4WebServices.API.Tests/Services/Authorization/PermissionServiceTestFixture.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            this.permissionService.ParticipantDao = this.participantDao.Object;
+
+            this.requirement = new Requirement(Guid.NewGuid(), 0);
+            this.definition = new Definition(Guid.NewGuid(), 0);
+
+            this.requirement.Definition.Add(this.requirement.Iid);
+        }
+
+        [Test]
+        public void VerifySameAsContainerPermissionAutorization()
```
Comment (samatrhea) [inline CDP4WebServices.API.Tests/Services/Authorization/PermissionServiceTestFixture.cs:134]: can you also add tests for:
  - a Thing that is contained by SiteDirectory
  - a Thing that is contained by EngineeringModel
  - a Thing that is contained by Iteration

**Review by samatrhea — CHANGES_REQUESTED** [review]
code looks good, just want to make sure we have 3 edge cases covered:
  - SiteDirectory
  - EngineeringModel
  - Iteration

**Comment by lxatrhea** [comment]
> 
> 
> So how do you resolve the container now? Not obvious from the code.

The injected `IResolveService` is used for that.

**Review by samatrhea — APPROVED** [review]

**Review by alexatrhea — APPROVED** [review]

**Pull request merged**